### PR TITLE
Async: GPIO

### PR DESCRIPTION
**Pull request by MabezDev**
- Upgrade to new eha
- Implement `Wait` for GpioPin
- Add an example for C3

## TODO
- [x] Add examples for all chips
- [x] Make the interrupt handler cleaner by being able to conjure a pin type

## Blocked on
- [x] We need `impl_trait_projections` compiler feature to be available in the Xtensa fork (26th of Jan )
- [x] #332 

## future work (probably not going in this PR)

- Adding support for multi-core chips that have interrupt registration per core.

### CODE REVIEW

File: esp-hal-common/src/gpio.rs [hunk 1]
```diff
@@ -1642,3 +1646,119 @@ pub(crate) use gpio;
...
+        Bank1GpioRegisterAccess::write_interrupt_status_clear(!0);
+
+        while intrs != 0 {
+            let pin_nr = intrs.trailing_zeros();
+            cfg_if::cfg_if! {
+                if #[cfg(any(esp32, esp32s2, esp32s3))] {
+                    if pin_nr < 32 {
+                        Bank0GpioRegisterAccess::set_int_enable(pin_nr as u8, 0, 0, false);
+                    } else {
+                        Bank0GpioRegisterAccess::set_int_enable(pin_nr as u8, 0, 0, false);
```
Comment (MabezDev) [inline esp-hal-common/src/gpio.rs:1754]: Oops, this should be `Bank0GpioRegisterAccess`

**Review by bjoernQ — APPROVED** [review]
LGTM

Only nit-pick is, there is no explanation of what to expect from the example - we have that for most examples - but not the end of the world and shouldn't keep us from merging

**Comment by MabezDev** [comment]
Good point, I added them for both this example and the embassy hello world :)

**Comment by MabezDev** [comment]
FYI what should we do about MSRV? 1.67 is only needed for async?

**Comment by bjoernQ** [comment]
> Btw what should we do about MSRV? 1.67 is only needed for async?

Maybe we can keep 1.65 and just say for async it's 1.67 I think I have seen that for other crates, too

**Review by jessebraham — APPROVED** [review]
LGTM!

RE the MSRV, I think what @bjoernQ said is probably best, keep it as-is but mention that for async specifically it's `1.67`. Once we've had another toolchain release or two we can re-visit this if needed.

File: esp-hal-common/src/gpio.rs [hunk 2]
```diff
@@ -1642,3 +1646,119 @@ pub(crate) use gpio;
...
+
+        let mut intrs = Bank0::pro_cpu_interrupt_status_read() as u64;
+
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        {
+            intrs |= (Bank1::pro_cpu_interrupt_status_read() as u64) << 32;
+        }
+
+        // clear interrupts
+        Bank0GpioRegisterAccess::write_interrupt_status_clear(!0);
```
Comment (bugadani) [inline esp-hal-common/src/gpio.rs:1743]: @MabezDev I'm not entirely sure but I think that `set_int_enable` does not implicitly clear interrupt status bits, it is possible for an edge-triggered listener to wake a task twice (in this call, and at a later GPIO interrupt request).

Comment (MabezDev) [reply esp-hal-common/src/gpio.rs:1743]: I don't quite understand your first point, but I do agree that we should only clear the interrupts read from the status inside the GPIO interrupt handler.

Comment (bugadani) [reply esp-hal-common/src/gpio.rs:1743]: I am in the process of opening an issue with more data that confirms my suspicions: level-triggered interrupts fire multiple times, it looks like.

**Pull request merged**